packages/react-dom/src/events/ReactBrowserEventEmitter.js

```diff
@@ -84,22 +85,46 @@ import isEventSupported from './isEventSupported';
  *    React Core     .  General Purpose Event Plugin System
  */
 
-const alreadyListeningTo = {};
-let reactTopListenersCounter = 0;
+const PossiblyWeakMap = typeof WeakMap === 'function' ? WeakMap : Map;
+const elementListeningObjects:
+  | WeakMap
+  | Map<
+      Document | Element | Node,
+      ElementListeningObject,
+    > = new PossiblyWeakMap();
 
-/**
- * To ensure no conflicts with other potential React instances on the page
- */
-const topListenersIDKey = '_reactListenersID' + ('' + Math.random()).slice(2);
+// This object will contain both legacy and non legacy events.
+// In the case of legacy events, we register an event listener
+// without passing an object with { passive, capture } etc to
+// third argument of the addEventListener call (we use a boolean).
+// For non legacy events, we register an event listener with
+// an object (third argument) with the passive property. We also
+// double listen for the event on the element, as we need to check
+// both paths (where passive is both true and false) so we can
+// handle logic in either case.
+export type ElementListeningObject = {
+  legacy: Set<DOMTopLevelEventType>,
+  nonLegacy: Set<DOMTopLevelEventType>,
+};
 
-function getListeningForDocument(mountAt: any) {
-  // In IE8, `mountAt` is a host object and doesn't have `hasOwnProperty`
-  // directly.
-  if (!Object.prototype.hasOwnProperty.call(mountAt, topListenersIDKey)) {
-    mountAt[topListenersIDKey] = reactTopListenersCounter++;
-    alreadyListeningTo[mountAt[topListenersIDKey]] = {};
+function createElementListeningObject(): ElementListeningObject {
+  return {
+    legacy: new Set(),
+    nonLegacy: new Set(),
+  };
+}
+
+function getListeningSetForElement(
+  element: Document | Element | Node,
+  isLegacy: boolean,
+): Set<DOMTopLevelEventType> {
+  let listeningObject = elementListeningObjects.get(element);
+  if (listeningObject === undefined) {
+    listeningObject = createElementListeningObject();
+    elementListeningObjects.set(element, listeningObject);
   }
-  return alreadyListeningTo[mountAt[topListenersIDKey]];
+  const listeningKey = isLegacy ? 'legacy' : 'nonLegacy';
+  return listeningObject[listeningKey];
```

Comment:
Can you make this if/else with two concrete names instead of access through a string?